graceful_controller_ros/src/graceful_controller_ros.cpp

```diff
@@ -384,134 +412,166 @@ bool GracefulControllerROS::computeVelocityCommands(geometry_msgs::Twist& cmd_ve
       break;
     }
 
-    // Configure controller max velocity based on current speed
-    if (!odom_helper_.getOdomTopic().empty())
+    // Iteratively try to find a path, incrementally reducing the velocity
+    double sim_velocity = max_vel_x;
+    do
     {
-      // The API of the OdometryHelperROS uses a PoseStamped
-      // but the data returned is velocities (should be a Twist)
-      geometry_msgs::PoseStamped robot_velocity;
-      odom_helper_.getRobotVel(robot_velocity);
-      double max_vel_x = robot_velocity.pose.position.x + (acc_lim_x_ * acc_dt_);
-      max_vel_x = std::min(max_vel_x, max_vel_x_);
-      max_vel_x = std::max(max_vel_x, min_vel_x_);
-      controller_->setVelocityLimits(min_vel_x_, max_vel_x, max_vel_theta_);
+      // Configure controller max velocity
+      controller_->setVelocityLimits(min_vel_x_, sim_velocity, max_vel_theta_);
+      // Actually simulate our path
+      if (simulate(target_pose, cmd_vel))
+      {
+        // Have valid command
+        return true;
+      }
+      // Reduce velocity and try again for same target_pose
+      sim_velocity -= scaling_step_;
```

Comment:
interesting, so we start at the highest velocity, and keep iterating until we are under scaling_vel_x_? Why are we using scaling_vel_x_ here instead of just min_vel_x?

Reply:
once you go below scaling_vel_x you are no longer shrinking the footprint (it's at scale = 1.0) - so if the path to the target_pose was blocked at the previous iteration, it will still be blocked at any lower velocity.